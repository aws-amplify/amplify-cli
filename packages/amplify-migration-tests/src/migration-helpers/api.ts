import {
  AddApiOptions,
  defaultOptions,
  getCLIPath,
  getSchemaPath,
  KEY_DOWN_ARROW,
  nspawn as spawn,
  selectRuntime,
  selectTemplate,
} from 'amplify-e2e-core';
import { assign } from 'lodash';

/**
 * Old Dx prior to this api workflow change https://github.com/aws-amplify/amplify-cli/pull/8153
 */
export function addApiWithoutSchemaOldDx(cwd: string, opts: Partial<AddApiOptions> = {}) {
  const options = assign(defaultOptions, opts);
  return new Promise<void>((resolve, reject) => {
    spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true })
      .wait('Please select from one of the below mentioned services:')
      .sendCarriageReturn()
      .wait('Provide API name:')
      .sendLine(options.apiName)
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendCarriageReturn()
      .wait(/.*Enter a description for the API key.*/)
      .sendCarriageReturn()
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendCarriageReturn()
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendCarriageReturn()
      .wait('Do you have an annotated GraphQL schema?')
      .sendConfirmNo()
      .wait('Choose a schema template:')
      .sendCarriageReturn()
      .wait('Do you want to edit the schema now?')
      .sendConfirmNo()
      .wait(
        '"amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud',
      )
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function addApiWithSchemaAndConflictDetectionOldDx(cwd: string, schemaFile: string) {
  const schemaPath = getSchemaPath(schemaFile);
  return new Promise<void>((resolve, reject) => {
    spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true })
      .wait('Please select from one of the below mentioned services:')
      .sendCarriageReturn()
      .wait('Provide API name:')
      .sendCarriageReturn()
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendCarriageReturn()
      .wait(/.*Enter a description for the API key.*/)
      .sendCarriageReturn()
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendCarriageReturn()
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendLine(KEY_DOWN_ARROW) // Down
      .wait(/.*Configure additional auth types.*/)
      .sendConfirmNo()
      .wait(/.*Enable conflict detection.*/)
      .sendConfirmYes()
      .wait(/.*Select the default resolution strategy.*/)
      .sendCarriageReturn()
      .wait(/.*Do you have an annotated GraphQL schema.*/)
      .sendConfirmYes()
      .wait('Provide your schema file path:')
      .sendLine(schemaPath)
      .wait(
        '"amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud',
      )
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function addRestApiOldDx(cwd: string, settings: any) {
  const isFirstRestApi = settings.isFirstRestApi ?? true;
  let chain = spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true })
    .wait('Select from one of the below mentioned services')
    .sendKeyDown()
    .sendCarriageReturn(); // REST

  if (!isFirstRestApi) {
    chain.wait('Would you like to add a new path to an existing REST API');

    if (settings.path) {
      chain
        .sendYes()
        .wait('Select the REST API you want to update')
        .sendCarriageReturn() // Select the first REST API
        .wait('What would you like to do?')
        .sendCarriageReturn() // Add another path
        .wait('Provide a path')
        .sendLine(settings.path)
        .wait('Choose a lambda source')
        .sendKeyDown()
        .sendCarriageReturn() // Existing lambda
        .wait('Choose the Lambda function to invoke by this path')
        .sendCarriageReturn() // Pick first one
        .wait('Restrict API access')
        .sendConfirmNo() // Do not restrict access
        .wait('Do you want to add another path')
        .sendConfirmNo() // Do not add another path
        .sendEof();

      return chain.runAsync();
    } else {
      chain.sendConfirmNo();
    }
  }

  chain.wait('Provide a friendly name for your resource to be used as a label for this category in the project');
  if (settings.apiName) {
    chain.sendLine(settings.apiName);
  } else {
    chain.sendCarriageReturn();
  }
  chain.wait('Provide a path').sendCarriageReturn().wait('Choose a lambda source');

  if (settings.existingLambda) {
    chain
      .sendKeyDown()
      .sendCarriageReturn() // Existing lambda
      .wait('Choose the Lambda function to invoke by this path'); // Expect only 1 Lambda is present
  } else {
    chain
      .sendCarriageReturn() // Create new Lambda function
      .wait('Provide an AWS Lambda function name')
      .sendCarriageReturn();

    selectRuntime(chain, 'nodejs');

    const templateName = settings.isCrud
      ? 'CRUD function for DynamoDB (Integration with API Gateway)'
      : 'Serverless ExpressJS function (Integration with API Gateway)';
    selectTemplate(chain, templateName, 'nodejs');

    if (settings.isCrud) {
      chain
        .wait('Choose a DynamoDB data source option')
        .sendCarriageReturn() // Use DDB table configured in current project
        .wait('Choose from one of the already configured DynamoDB tables')
        .sendCarriageReturn(); // Use first one in the list
    }

    chain
      .wait('Do you want to configure advanced settings?')
      .sendConfirmNo()
      .wait('Do you want to edit the local lambda function now')
      .sendConfirmNo();
  }

  chain.wait('Restrict API access');
  if (settings.restrictAccess) {
    chain.sendConfirmYes();

    if (settings.hasUserPoolGroups) {
      chain.wait('Restrict access by').sendCarriageReturn(); // Auth/Guest Users
    }

    chain.wait('Who should have access');

    if (settings.allowGuestUsers) {
      chain
        .sendKeyDown()
        .sendCarriageReturn() // Authenticated and Guest users
        .wait('What permissions do you want to grant to Authenticated users')
        .send('a') // CRUD permissions for authenticated users
        .sendCarriageReturn()
        .wait('What permissions do you want to grant to Guest users')
        .send('a') // CRUD permissions for guest users
        .sendCarriageReturn();
    } else {
      chain
        .sendCarriageReturn() // Authenticated users only
        .wait('What permissions do you want to grant to Authenticated users')
        .send('a') // CRUD permissions
        .sendCarriageReturn();
    }
  } else {
    chain.sendConfirmNo(); // Do not restrict access
  }

  chain.wait('Do you want to add another path').sendConfirmNo().sendEof();

  return chain.runAsync();
}
