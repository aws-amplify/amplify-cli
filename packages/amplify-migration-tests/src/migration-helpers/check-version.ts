import { getCLIPath, nspawn as spawn } from 'amplify-e2e-core';

export function versionCheck(cwd: string, testingWithLatestCodebase = false, version: { v?: string } = {}): Promise<void> {
  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['-v'], { cwd, stripColors: true })
      .wait(/\d+\.\d+\.\d+/, v => (version.v = v.trim()))
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

/**
 * This list is used to check migration tests with the following changes. (excludes layer migration tests)
 *
 * api add/update flow: https://github.com/aws-amplify/amplify-cli/pull/8287
 *
 * ext migrate flow: https://github.com/aws-amplify/amplify-cli/pull/8806
 */
export const allowedVersionsToMigrateFrom = ['5.2.0', '6.1.0'];
