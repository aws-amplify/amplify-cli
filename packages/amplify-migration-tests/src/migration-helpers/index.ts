export * from './init';
export * from './api';
export * from './check-version';
export * from './legacy-lambda-layer';
