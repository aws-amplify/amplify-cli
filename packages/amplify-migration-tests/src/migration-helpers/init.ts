import { addCircleCITags, getCLIPath, nspawn as spawn } from 'amplify-e2e-core';
import { EOL } from 'os';

const defaultSettings = {
  name: EOL,
  envName: 'integtest',
  editor: EOL,
  appType: EOL,
  framework: EOL,
  srcDir: EOL,
  distDir: EOL,
  buildCmd: EOL,
  startCmd: EOL,
  useProfile: EOL,
  profileName: EOL,
  region: process.env.CLI_REGION,
  local: false,
  disableAmplifyAppCreation: true,
};

export function initJSProjectWithProfileOldDX(cwd: string, settings: Object, testingWithLatestCodebase = false): Promise<void> {
  const s = { ...defaultSettings, ...settings };
  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  addCircleCITags(cwd);

  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['init'], { cwd, stripColors: true, env })
      .wait('Enter a name for the project')
      .sendLine(s.name)
      .wait('Enter a name for the environment')
      .sendLine(s.envName)
      .wait('Choose your default editor:')
      .sendLine(s.editor)
      .wait("Choose the type of app that you're building")
      .sendLine(s.appType)
      .wait('What javascript framework are you using')
      .sendLine(s.framework)
      .wait('Source Directory Path:')
      .sendLine(s.srcDir)
      .wait('Distribution Directory Path:')
      .sendLine(s.distDir)
      .wait('Build Command:')
      .sendLine(s.buildCmd)
      .wait('Start Command:')
      .sendCarriageReturn()
      .wait('Using default provider  awscloudformation')
      .wait(/(Select the authentication method you want to use|Do you want to use an AWS profile)/)
      .sendCarriageReturn()
      .wait('Please choose the profile you want to use')
      .sendLine(s.profileName)
      .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function initJSProjectWithProfile(cwd: string, settings: Object, testingWithLatestCodebase = false): Promise<void> {
  const s = { ...defaultSettings, ...settings };
  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  addCircleCITags(cwd);

  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['init'], { cwd, stripColors: true, env })
      .wait('Enter a name for the project')
      .sendLine(s.name)
      .wait('Initialize the project with the above configuration?')
      .sendConfirmNo()
      .wait('Enter a name for the environment')
      .sendLine(s.envName)
      .wait('Choose your default editor:')
      .sendLine(s.editor)
      .wait("Choose the type of app that you're building")
      .sendLine(s.appType)
      .wait('What javascript framework are you using')
      .sendLine(s.framework)
      .wait('Source Directory Path:')
      .sendLine(s.srcDir)
      .wait('Distribution Directory Path:')
      .sendLine(s.distDir)
      .wait('Build Command:')
      .sendLine(s.buildCmd)
      .wait('Start Command:')
      .sendCarriageReturn()
      .wait('Using default provider  awscloudformation')
      .wait(/(Select the authentication method you want to use|Do you want to use an AWS profile)/)
      .sendCarriageReturn()
      .wait('Please choose the profile you want to use')
      .sendLine(s.profileName)
      .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}
