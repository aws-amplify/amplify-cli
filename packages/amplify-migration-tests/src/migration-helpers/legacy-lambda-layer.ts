import { pathManager } from 'amplify-cli-core';
import {
  ExecutionContext,
  getCLIPath,
  getRuntimeDisplayNames,
  LayerPermissionChoice,
  LayerRuntime,
  layerRuntimeChoices,
  multiSelect,
  nspawn as spawn,
  permissionChoices,
} from 'amplify-e2e-core';
import * as fs from 'fs';
import * as path from 'path';

export function legacyAddLayer(
  cwd: string,
  settings: {
    layerName: string;
    permissions?: LayerPermissionChoice[];
    accountId?: string;
    orgId?: string;
    runtimes: LayerRuntime[];
  },
): Promise<void> {
  const defaultSettings = {
    permissions: [],
  };
  settings = { ...defaultSettings, ...settings };
  return new Promise((resolve, reject) => {
    const chain: ExecutionContext = spawn(getCLIPath(false), ['add', 'function'], { cwd, stripColors: true })
      .wait('Select which capability you want to add:')
      .sendKeyDown()
      .sendCarriageReturn() // Layer
      .wait('Provide a name for your Lambda layer:')
      .sendLine(settings.layerName);

    const runtimeDisplayNames = getRuntimeDisplayNames(settings.runtimes);
    expect(settings.runtimes.length === runtimeDisplayNames.length).toBe(true);

    chain.wait('Select up to 2 compatible runtimes:');
    multiSelect(chain, runtimeDisplayNames, layerRuntimeChoices);
    chain.wait('The current AWS account will always have access to this layer.');

    multiSelect(chain, settings.permissions, permissionChoices);

    if (settings.permissions.includes('Specific AWS accounts')) {
      chain.wait('Provide a list of comma-separated AWS account IDs:').sendLine(settings.accountId);
    }

    if (settings.permissions.includes('Specific AWS organization')) {
      chain.wait('Provide a list of comma-separated AWS organization IDs:').sendLine(settings.orgId);
    }

    if (settings.runtimes.length > 0) {
      chain.wait('Move your libraries to the following folder:');
    }

    chain.run((err: Error) => {
      if (!err) {
        resolve();
      } else {
        reject(err);
      }
    });
  });
}

export function legacyAddOptData(projRoot: string, layerName: string): void {
  fs.writeFileSync(path.join(projRoot, 'amplify', 'backend', 'function', layerName, 'opt', 'data.txt'), 'data', 'utf8');
}

export function legacyUpdateOptData(projRoot: string, layerName: string, data: string): void {
  fs.writeFileSync(path.join(projRoot, 'amplify', 'backend', 'function', layerName, 'opt', 'data.txt'), data, 'utf8');
}

export function validateLayerConfigFilesMigrated(projRoot: string, layerName: string) {
  const layerDirPath = pathManager.getResourceDirectoryPath(projRoot, 'function', layerName);
  return (
    fs.existsSync(path.join(layerDirPath, 'layer-configuration.json')) &&
    !fs.existsSync(path.join(layerDirPath, 'layer-runtimes.json')) &&
    !fs.existsSync(path.join(layerDirPath, 'layer-parameters.json'))
  );
}
