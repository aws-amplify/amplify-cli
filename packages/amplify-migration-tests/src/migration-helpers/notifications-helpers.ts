import { singleSelect, getCLIPath, nspawn as spawn } from '@aws-amplify/amplify-e2e-core';

/**
 * Adds notification resource for a given channel
 */
export const addLegacySmsNotificationChannel = async (cwd: string, resourceName: string, hasAnalytics = false): Promise<void> => {
  const chain = spawn(getCLIPath(false), ['add', 'notification'], { cwd, stripColors: true });

  singleSelect(chain.wait('Choose the push notification channel to enable'), 'SMS', ['APNS', 'FCM', 'Email', 'SMS']);

  if (!hasAnalytics) {
    chain.wait('Provide your pinpoint resource name').sendLine(resourceName);
  }

  return chain.wait(`The SMS channel has been successfully enabled`).sendEof().runAsync();
};

/**
 * Removes all notifications channels
 */
export const removeLegacyAllNotificationChannel = async (cwd: string): Promise<void> =>
  spawn(getCLIPath(false), ['remove', 'notifications'], { cwd, stripColors: true })
    .wait('Choose what to remove. (Use arrow keys)')
    .sendKeyDown() // The Pinpoint application
    .sendCarriageReturn()
    .wait(`Confirm that you want to delete the associated Amazon Pinpoint application`)
    .sendConfirmYes()
    .sendEof()
    .runAsync();
