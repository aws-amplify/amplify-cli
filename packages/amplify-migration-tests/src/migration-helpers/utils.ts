import {
  amplifyPull,
  amplifyPushForce,
  amplifyPushWithoutCodegen,
  cliInputsExists,
  createNewProjectDir,
  deleteProjectDir,
  getAppId,
  getBackendConfig,
  getCLIInputs,
  getCloudFormationTemplate,
  getParameters,
  parametersExists,
} from '@aws-amplify/amplify-e2e-core';
import * as cfnDiff from '@aws-cdk/cloudformation-diff';
import { Writable } from 'stream';
import { AmplifyCategories } from '@aws-amplify/amplify-cli-core';
import strip from 'strip-ansi';
import { v4 as uuid } from 'uuid';

/**
 * generates a random string
 */
export const getShortId = (): string => {
  const [shortId] = uuid().split('-');

  return shortId;
};

/**
 * Given the parameter objects of each project for the provided category & resource key,
 * you can modify the parameter objects before the diff is performed.
 *
 * You can use conditional logic to delete attributes on the parameter objects before they are diffed, if
 * you want to exclude those attributes from the comparison. Make sure to return the modified objects
 * that you want to be diffed.
 */
export type ExcludeFromParameterDiff = (
  currentCategory: string,
  currentResourceKey: string,
  parameters: {
    project1: Record<string, unknown>;
    project2: Record<string, unknown>;
  },
) => { project1: Record<string, unknown>; project2: Record<string, unknown> };

/**
 * Asserts that parameters between two project directories didn't drift.
 */
export const assertNoParameterChangesBetweenProjects = (
  projectRoot1: string,
  projectRoot2: string,
  options?: {
    excludeFromParameterDiff?: ExcludeFromParameterDiff;
  },
): void => {
  const backendConfig1 = getBackendConfig(projectRoot1);
  const backendConfig2 = getBackendConfig(projectRoot2);
  expect(backendConfig2).toMatchObject(backendConfig1);
  for (const categoryKey of Object.keys(backendConfig1)) {
    const category = backendConfig1[categoryKey];
    if (Object.values(AmplifyCategories).includes(categoryKey)) {
      for (const resourceKey of Object.keys(category)) {
        if (cliInputsExists(projectRoot1, categoryKey, resourceKey)) {
          const cliInputs1 = getCLIInputs(projectRoot1, categoryKey, resourceKey);
          const cliInputs2 = getCLIInputs(projectRoot2, categoryKey, resourceKey);
          expect(cliInputs1).toEqual(cliInputs2);
        }
        if (parametersExists(projectRoot1, categoryKey, resourceKey)) {
          let parameters1 = getParameters(projectRoot1, categoryKey, resourceKey);
          let parameters2 = getParameters(projectRoot2, categoryKey, resourceKey);
          if (options && options.excludeFromParameterDiff) {
            const afterExclusions = options.excludeFromParameterDiff(categoryKey, resourceKey, {
              project1: parameters1,
              project2: parameters2,
            });
            parameters1 = afterExclusions.project1;
            parameters2 = afterExclusions.project2;
          }
          expect(parameters1).toEqual(parameters2);
        }
      }
    }
  }
};

class InMemoryWritable extends Writable {
  private _payload = '';

  _write(chunk: unknown, __encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
    if (chunk) {
      this._payload += chunk.toString();
    }
    callback();
  }

  toString(): string {
    return this._payload;
  }
}

/**
 * Given the CFN templates of each project for the provided category & resource key,
 * you can modify the CFN templates objects before the diff is performed.
 *
 * You can use conditional logic to delete attributes on the CFN object before they are diffed, if
 * you want to exclude those attributes from the comparison. Make sure to return the modified objects
 * that you want to be diffed.
 */
export type ExcludeFromCFNDiff = (
  currentCategory: string,
  currentResourceKey: string,
  cfnTemplates: {
    project1: Record<string, unknown>;
    project2: Record<string, unknown>;
  },
) => { project1: Record<string, unknown>; project2: Record<string, unknown> };

/**
 * Collects all differences between cloud formation templates into a single string.
 */
export const collectCloudformationDiffBetweenProjects = (
  projectRoot1: string,
  projectRoot2: string,
  excludeFn?: ExcludeFromCFNDiff,
): string => {
  const backendConfig1 = getBackendConfig(projectRoot1);
  const backendConfig2 = getBackendConfig(projectRoot2);
  expect(backendConfig2).toMatchObject(backendConfig1);
  const stream = new InMemoryWritable();
  for (const categoryKey of Object.keys(backendConfig1)) {
    const category = backendConfig1[categoryKey];
    for (const resourceKey of Object.keys(category)) {
      let template1 = getCloudFormationTemplate(projectRoot1, categoryKey, resourceKey);
      let template2 = getCloudFormationTemplate(projectRoot2, categoryKey, resourceKey);

      // Description does not matter much and it can contain os/runtime specific words.
      delete template1.Description;
      delete template2.Description;

      if (excludeFn) {
        const afterExclusions = excludeFn(categoryKey, resourceKey, { project1: template1, project2: template2 });
        template1 = afterExclusions.project1;
        template2 = afterExclusions.project2;
      }

      const templateDiff = cfnDiff.diffTemplate(template1, template2);
      if (!templateDiff.isEmpty) {
        cfnDiff.formatDifferences(stream, templateDiff);
      }
    }
  }
  return strip(stream.toString());
};

/**
 * Pulls and pushes project with latest codebase. Validates parameter and cfn drift.
 */
export const pullPushWithLatestCodebaseValidateParameterAndCfnDrift = async (projRoot: string, projName: string): Promise<void> => {
  const appId = getAppId(projRoot);
  expect(appId).toBeDefined();
  const projRoot2 = await createNewProjectDir(`${projName}2`);
  try {
    await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
    await amplifyPushWithoutCodegen(projRoot2, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
  } finally {
    deleteProjectDir(projRoot2);
  }
};

/**
 * Pulls and pushes with force project with latest codebase. Validates parameter and cfn drift.
 */
export const pullPushForceWithLatestCodebaseValidateParameterAndCfnDrift = async (projRoot: string, projRoot2: string): Promise<void> => {
  const appId = getAppId(projRoot);
  expect(appId).toBeDefined();
  try {
    await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
    await amplifyPushForce(projRoot2, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
  } finally {
    deleteProjectDir(projRoot2);
  }
};
