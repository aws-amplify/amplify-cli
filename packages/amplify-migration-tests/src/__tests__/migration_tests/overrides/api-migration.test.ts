import {
  addHeadlessApi,
  amplifyPush,
  amplifyPushUpdate,
  createNewProjectDir,
  deleteProject,
  deleteProjectDir,
  getAppSyncApi,
  getCLIInputs,
  getProjectConfig,
  getProjectMeta,
  getProjectSchema,
  getSchemaPath,
  getTransformConfig,
  initJSProjectWithProfile,
  updateApiSchema,
  updateApiWithMultiAuth,
  updateAPIWithResolutionStrategyWithModels,
  updateHeadlessApi,
} from 'amplify-e2e-core';
import { AddApiRequest, UpdateApiRequest } from 'amplify-headless-interface';
import * as fs from 'fs-extra';
import { TRANSFORM_BASE_VERSION, TRANSFORM_CURRENT_VERSION } from 'graphql-transformer-core';
import { join } from 'path';
import { addApiWithoutSchemaOldDx, addApiWithSchemaAndConflictDetectionOldDx } from '../../../migration-helpers';

describe('api migration update test', () => {
  let projRoot: string;
  beforeEach(async () => {
    projRoot = await createNewProjectDir('graphql-api');
  });

  afterEach(async () => {
    const metaFilePath = join(projRoot, 'amplify', '#current-cloud-backend', 'amplify-meta.json');
    if (fs.existsSync(metaFilePath)) {
      await deleteProject(projRoot, undefined, true);
    }
    deleteProjectDir(projRoot);
  });

  it('api update migration with multiauth', async () => {
    // init and add api with installed CLI
    await initJSProjectWithProfile(projRoot, { name: 'simplemodelmultiauth' });
    await addApiWithoutSchemaOldDx(projRoot);
    await updateApiSchema(projRoot, 'simplemodelmultiauth', 'simple_model.graphql');
    await amplifyPush(projRoot);
    // update and push with codebase
    await updateApiWithMultiAuth(projRoot, { testingWithLatestCodebase: true });
    // cli-inputs should exist
    expect(getCLIInputs(projRoot, 'api', 'simplemodelmultiauth')).toBeDefined();
    await amplifyPushUpdate(projRoot, undefined, true, true);

    const meta = getProjectMeta(projRoot);
    const { output } = meta.api.simplemodelmultiauth;
    const { GraphQLAPIIdOutput, GraphQLAPIEndpointOutput, GraphQLAPIKeyOutput } = output;
    const { graphqlApi } = await getAppSyncApi(GraphQLAPIIdOutput, meta.providers.awscloudformation.Region);

    expect(graphqlApi).toBeDefined();
    expect(graphqlApi.authenticationType).toEqual('API_KEY');
    expect(graphqlApi.additionalAuthenticationProviders).toHaveLength(3);
    expect(graphqlApi.additionalAuthenticationProviders).toHaveLength(3);

    const cognito = graphqlApi.additionalAuthenticationProviders.filter(a => a.authenticationType === 'AMAZON_COGNITO_USER_POOLS')[0];

    expect(cognito).toBeDefined();
    expect(cognito.userPoolConfig).toBeDefined();

    const iam = graphqlApi.additionalAuthenticationProviders.filter(a => a.authenticationType === 'AWS_IAM')[0];

    expect(iam).toBeDefined();

    const oidc = graphqlApi.additionalAuthenticationProviders.filter(a => a.authenticationType === 'OPENID_CONNECT')[0];

    expect(oidc).toBeDefined();
    expect(oidc.openIDConnectConfig).toBeDefined();
    expect(oidc.openIDConnectConfig.issuer).toEqual('https://facebook.com/');
    expect(oidc.openIDConnectConfig.clientId).toEqual('clientId');
    expect(oidc.openIDConnectConfig.iatTTL).toEqual(1000);
    expect(oidc.openIDConnectConfig.authTTL).toEqual(2000);

    expect(GraphQLAPIIdOutput).toBeDefined();
    expect(GraphQLAPIEndpointOutput).toBeDefined();
    expect(GraphQLAPIKeyOutput).toBeDefined();

    expect(graphqlApi).toBeDefined();
    expect(graphqlApi.apiId).toEqual(GraphQLAPIIdOutput);
  });

  it('init a sync enabled project and update conflict resolution strategy', async () => {
    const name = `syncenabled`;
    // init and add api with locally installed cli
    await initJSProjectWithProfile(projRoot, { name });
    await addApiWithSchemaAndConflictDetectionOldDx(projRoot, 'simple_model.graphql');
    await amplifyPush(projRoot);
    let transformConfig = getTransformConfig(projRoot, name);
    expect(transformConfig).toBeDefined();
    expect(transformConfig.ResolverConfig).toBeDefined();
    expect(transformConfig.ResolverConfig.project).toBeDefined();
    expect(transformConfig.ResolverConfig.project.ConflictDetection).toEqual('VERSION');
    expect(transformConfig.ResolverConfig.project.ConflictHandler).toEqual('AUTOMERGE');

    //update and push with codebase
    await updateAPIWithResolutionStrategyWithModels(projRoot, { testingWithLatestCodebase: true });
    expect(getCLIInputs(projRoot, 'api', 'syncenabled')).toBeDefined();
    transformConfig = getTransformConfig(projRoot, name);
    expect(transformConfig).toBeDefined();
    expect(transformConfig.Version).toBeDefined();
    expect(transformConfig.Version).toEqual(TRANSFORM_CURRENT_VERSION);
    expect(transformConfig.ResolverConfig).toBeDefined();
    expect(transformConfig.ResolverConfig.project).toBeDefined();
    expect(transformConfig.ResolverConfig.project.ConflictDetection).toEqual('VERSION');
    expect(transformConfig.ResolverConfig.project.ConflictHandler).toEqual('OPTIMISTIC_CONCURRENCY');

    await amplifyPushUpdate(projRoot, undefined, true, true);
    const meta = getProjectMeta(projRoot);
    const { output } = meta.api[name];
    const { GraphQLAPIIdOutput, GraphQLAPIEndpointOutput, GraphQLAPIKeyOutput } = output;
    const { graphqlApi } = await getAppSyncApi(GraphQLAPIIdOutput, meta.providers.awscloudformation.Region);

    expect(GraphQLAPIIdOutput).toBeDefined();
    expect(GraphQLAPIEndpointOutput).toBeDefined();
    expect(GraphQLAPIKeyOutput).toBeDefined();

    expect(graphqlApi).toBeDefined();
    expect(graphqlApi.apiId).toEqual(GraphQLAPIIdOutput);
  });

  const addApiRequest: AddApiRequest = {
    version: 1,
    serviceConfiguration: {
      serviceName: 'AppSync',
      apiName: 'myApiName',
      transformSchema: fs.readFileSync(getSchemaPath('simple_model.graphql'), 'utf8'),
      defaultAuthType: {
        mode: 'API_KEY',
      },
    },
  };

  const updateApiRequest: UpdateApiRequest = {
    version: 1,
    serviceModification: {
      serviceName: 'AppSync',
      transformSchema: fs.readFileSync(getSchemaPath('simple_model_override.graphql'), 'utf8'),
      defaultAuthType: {
        mode: 'AWS_IAM',
      },
      additionalAuthTypes: [
        {
          mode: 'API_KEY',
        },
      ],
      conflictResolution: {
        defaultResolutionStrategy: {
          type: 'OPTIMISTIC_CONCURRENCY',
        },
      },
    },
  };
  it('updates AppSync API in headless mode', async () => {
    const name = `simplemodelv${TRANSFORM_BASE_VERSION}`;
    await initJSProjectWithProfile(projRoot, {});
    await addHeadlessApi(projRoot, addApiRequest, {
      allowDestructiveUpdates: false,
      testingWithLatestCodebase: false,
    });
    await amplifyPush(projRoot);
    await updateHeadlessApi(projRoot, updateApiRequest, true);
    expect(getCLIInputs(projRoot, 'api', 'myApiName')).toBeDefined();
    await amplifyPushUpdate(projRoot, undefined, undefined, true);

    //verify
    const meta = getProjectMeta(projRoot);
    const { output } = meta.api.myApiName;
    const { GraphQLAPIIdOutput, GraphQLAPIEndpointOutput, GraphQLAPIKeyOutput } = output;
    const { graphqlApi } = await getAppSyncApi(GraphQLAPIIdOutput, meta.providers.awscloudformation.Region);

    expect(GraphQLAPIIdOutput).toBeDefined();
    expect(GraphQLAPIEndpointOutput).toBeDefined();
    expect(GraphQLAPIKeyOutput).toBeDefined();

    expect(graphqlApi).toBeDefined();
    expect(graphqlApi.apiId).toEqual(GraphQLAPIIdOutput);

    expect(getTransformConfig(projRoot, 'myApiName')).toMatchSnapshot();
    expect(output.authConfig).toMatchSnapshot();
    expect(getProjectSchema(projRoot, 'myApiName')).toMatchSnapshot();
  });
});
