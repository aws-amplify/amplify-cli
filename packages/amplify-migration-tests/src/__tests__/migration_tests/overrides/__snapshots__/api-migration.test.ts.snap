// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api migration update test updates AppSync API in headless mode 1`] = `
Object {
  "ElasticsearchWarning": true,
  "ResolverConfig": Object {
    "project": Object {
      "ConflictDetection": "VERSION",
      "ConflictHandler": "OPTIMISTIC_CONCURRENCY",
    },
  },
  "Version": 5,
}
`;

exports[`api migration update test updates AppSync API in headless mode 2`] = `
Object {
  "additionalAuthenticationProviders": Array [
    Object {
      "apiKeyConfig": Object {},
      "authenticationType": "API_KEY",
    },
  ],
  "defaultAuthentication": Object {
    "authenticationType": "AWS_IAM",
  },
}
`;

exports[`api migration update test updates AppSync API in headless mode 3`] = `
"type Todo @model {
  id: ID!
  content: String
  override: String
}
"
`;
